Mathlib/Data/Matroid/Closure.lean

```diff
@@ -0,0 +1,199 @@
+/-
+Copyright (c) 2024 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import Mathlib.Data.Matroid.Basic
+
+/-!
+# Matroid Closure
+
+A `Flat` of a matroid `M` is a combinatorial analogue of a subspace of a vector space,
+and is defined to be a subset `F` of the ground set of `M` such that for each basis
+`I` for `M`, every set having `I` as a basis is contained in `F`.
+
+The *closure* of a set `X` in a matroid `M` is the intersection of all flats of `M` containing `X`.
+This is a combinatorial analogue of the linear span of a set of vectors.
+
+For `M : Matroid α`, this file defines a predicate `M.Flat : Set α → Prop` and a function
+`M.cl : Set α → Set α` corresponding to these notions, and develops API for the latter.
+API for `Matroid.Flat` will appear in another file; we include the definition here since
+it is used in the definition of `Matroid.cl`.
+
+## Main definitions
+
+* For `M : Matroid α` and `F : Set α`, `M.Flat F` means that `F` is a flat of `M`.
+* For `M : Matroid α` and `X : Set α`, `M.cl X` is the closure of `X` in `M`.
+
+## Implementation details
+
+If `X : Set α` satisfies `X ⊆ M.E`, then it is clear how `M.cl X` should be defined.
+But `M.cl X` needs to be defined for all `X : Set α`, so a convention is needed for how `M.cl`
+handles sets containing junk elements outside `M.E`. Ideally, this convention should minimize
+the need to keep track of which sets are actually contained in the ground set.
+
+All such choices come with tradeoffs; the definition we opt for satisfies `M.cl X = M.cl (X ∩ M.E)`.
+The drawback here is that the statement `X ⊆ M.cl X` requires the assumption `X ⊆ M.E`.
+But all the other properties of closure work nicely without extra assumptions;
+closure is monotone, idempotent, and `M.cl X ⊆ M.E` for all `X`.
```

Comment:
Thinking out loud for a bit...

If you choose to add `X \ M.E` to the closure, then it is still monotone and idempotent. You get `X ⊆ M.cl X` without assumptions, but now `M.cl X ⊆ M.E` will need an assumption.

On the other hand, it makes `M.cl` an example of a [ClosureOperator](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Closure.html#ClosureOperator). That would give you some API for free.

Naive question: is `M.cl X` an example of a flat? If so, I suspect that `M.cl` is part of a Galois correspondence (even Galois insertion) between flats and subsets. That would also give you a bunch of API for free.

Happy to hear your thoughts.

Reply:
Thanks for this - I wasn't familiar with the `GaloisConnection` and `ClosureOperator` parts of mathlib, and gave this quite a lot of thought. 

Mathematically, you're absolutely on the money - if we have `M.E = univ`, then `M.closure` would give a `ClosureOperator` on `Set α` for which `Flat ↔ IsClosed`, and we thus have a Galois insertion with the types `Set α` and `{F // M.Flat F}`. 

But because we may have `M.E ⊂ univ` and `IsClosed univ` is always true, we can't have all three of the following at once : 
(1) `M.E ⊂ univ`,
(2) every `Flat` is a subset of the ground set,
(3) `M.closure` is a `ClosureOperator` on `Set α` for which `Flat ↔ IsClosed`

For instance, if we included `X \ M.E` in `M.cl X`, then we would get a `ClosureOperator`, but `IsClosed` would not correspond to `Matroid.Flat`, since (for example) the set `univ` would be closed but not a flat. Perhaps having this `ClosureOperator` would still be somewhat useful, but losing the property that `M.cl X ⊆ M.E` is not great for `aesop_mat`, and I think in practice would add a lot of extra assumptions to lemmas.  

I believe that (1) and (2) are important enough for the design of matroids that we have to give up (3). 

However, when it comes to piggybacking off the `ClosureOperator` and `GaloisInsertion` API, we can do something using subtypes. 
See `Matroid.subtypeClosure` as defined below. It allowed me to golf the proof of `closure_iUnion_closure_eq_closure_iUnion`,
and I think it'll be nice for making the `Flat` API simpler when we get there. 

That all said, after pushing, I will play around with including `X \ M.E` in `M.cl X` to see what it entails in terms of extra assumptions.

Reply:
Thanks for writing down these reflections! Would you mind incorporating (parts of) this comment into the "Implementation notes" in the module docstring?

After that (and once CI passes), I think this PR is ready to be merged.

Reply:
Great, thanks. At the moment I'm experimenting with the `X ⊆ M.cl X` design to see how much havoc it wreaks in my larger repo. If that turns out to be 'a lot', I'll go ahead with what you say.

Reply:
One other option is to have both, where the current version would be consider "canonical" and the other is only there as implementation detail for setting up API. Because I do think that the other would give you a closure operator and a galois correspondence.
It could even be there only for illustrative purposes.